src/mlpack/tests/lmnn_test.cpp

```diff
@@ -0,0 +1,352 @@
+  arma::Row<size_t> labels = " 0    0    0    1    1    1   ";
+
+  LMNNFunction<> lmnnfn(dataset, labels, 1, 0.6);
+
+  arma::mat coordinates = arma::eye<arma::mat>(2, 2);
+  arma::mat gradient(2, 2);
+
+  double objective = lmnnfn.EvaluateWithGradient(coordinates, 0, gradient, 1);
+
+  BOOST_REQUIRE_CLOSE(objective, 1.576, 1e-5);
+
+  BOOST_REQUIRE_CLOSE(gradient(0, 0), -0.024, 1e-5);
+  BOOST_REQUIRE_CLOSE(gradient(0, 1), 0.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(gradient(1, 0), 0.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(gradient(1, 1), 1.0, 1e-5);
+
+  objective = lmnnfn.EvaluateWithGradient(coordinates, 1, gradient, 1);
+
+  BOOST_REQUIRE_CLOSE(objective, 1.576, 1e-5);
+
+  BOOST_REQUIRE_CLOSE(gradient(0, 0), -0.024, 1e-5);
+  BOOST_REQUIRE_CLOSE(gradient(0, 1), 0.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(gradient(1, 0), 0.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(gradient(1, 1), 1.0, 1e-5);
+
+  objective = lmnnfn.EvaluateWithGradient(coordinates, 2, gradient, 1);
+
+  BOOST_REQUIRE_CLOSE(objective, 1.576, 1e-5);
+
+  BOOST_REQUIRE_CLOSE(gradient(0, 0), -0.024, 1e-5);
+  BOOST_REQUIRE_CLOSE(gradient(0, 1), 0.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(gradient(1, 0), 0.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(gradient(1, 1), 1.0, 1e-5);
+
+  objective = lmnnfn.EvaluateWithGradient(coordinates, 3, gradient, 1);
+
+  BOOST_REQUIRE_CLOSE(objective, 1.576, 1e-5);
+
+  BOOST_REQUIRE_CLOSE(gradient(0, 0), -0.024, 1e-5);
+  BOOST_REQUIRE_CLOSE(gradient(0, 1), 0.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(gradient(1, 0), 0.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(gradient(1, 1), 1.0, 1e-5);
+
+  objective = lmnnfn.EvaluateWithGradient(coordinates, 4, gradient, 1);
+
+  BOOST_REQUIRE_CLOSE(objective, 1.576, 1e-5);
+
+  BOOST_REQUIRE_CLOSE(gradient(0, 0), -0.024, 1e-5);
+  BOOST_REQUIRE_CLOSE(gradient(0, 1), 0.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(gradient(1, 0), 0.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(gradient(1, 1), 1.0, 1e-5);
+
+  objective = lmnnfn.EvaluateWithGradient(coordinates, 5, gradient, 1);
+
+  BOOST_REQUIRE_CLOSE(objective, 1.576, 1e-5);
+
+  BOOST_REQUIRE_CLOSE(gradient(0, 0), -0.024, 1e-5);
+  BOOST_REQUIRE_CLOSE(gradient(0, 1), 0.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(gradient(1, 0), 0.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(gradient(1, 1), 1.0, 1e-5);
+}
+
+BOOST_AUTO_TEST_CASE(LMNNLBFGSSimpleDataset)
+{
+  // Useful but simple dataset with six points and two classes.
+  arma::mat dataset        = "-0.1 -0.1 -0.1  0.1  0.1  0.1;"
+                             " 1.0  0.0 -1.0  1.0  0.0 -1.0 ";
+  arma::Row<size_t> labels = " 0    0    0    1    1    1   ";
+
+  // Huge learning rate because this is so simple.
+  LMNN<SquaredEuclideanDistance, L_BFGS> lmnn(dataset, labels, 1);
+
+  arma::mat outputMatrix;
+  lmnn.LearnDistance(outputMatrix);
+
+  // Ensure that the objective function is better now.
+  LMNNFunction<> lmnnfn(dataset, labels, 1, 0.6);
+
+  double initObj = lmnnfn.Evaluate(arma::eye<arma::mat>(2, 2));
+  double finalObj = lmnnfn.Evaluate(outputMatrix);
```

Comment:
Hmm, this checks that the objective function is better, which is correlated to the result we want, but what we'd really like to see is that the nearest neighbor of each point now has the same label.  So maybe we should check that explicitly?

Reply:
I think, "final nearest neighbors having same labels" shouldn't be necessary condition. As that would be the ideal case. What do you think? 
Maybe we can do that check for this dataset explicitly.

Reply:
Oops, didn't respond to this one, sorry about that.  I think maybe it is worth checking that KNN accuracy improves for this very simple dataset, but you are right, in general that may not always be the case.

Reply:
I made a run on this dataset. We are getting 100% accuracy with this.

Reply:
Sounds good, we should write that into the test either by manual distance calculations or using the KNN object.